contracts/root/WithdrawManager.sol

```diff
@@ -0,0 +1,484 @@
+  * @param _utxoPos Position of the UTXO in the chain.
+  * @return A tuple representing the active exit for the given UTXO.
+  */
+  function getExit(uint256 _utxoPos)
+    public
+    view
+    returns (address, uint256)
+  {
+    return (exits[_utxoPos].token, exits[_utxoPos].amount);
+  }
+
+  /**
+  * @dev Determines the next exit to be processed.
+  * @param _token Asset type to be exited.
+  * @return A tuple of the position and time when this exit can be processed.
+  */
+  function getNextExit(address _token)
+    public
+    view
+    returns (uint256, uint256)
+  {
+    return PriorityQueue(exitsQueues[_token]).getMin();
+  }
+
+  /**
+  * @dev Processes any exits that have completed the exit period.
+  */
+  function processExits(address _token) public {
+    uint256 exitableAt;
+    uint256 utxoPos;
+
+    // retrieve priority queue
+    PriorityQueue exitQueue = PriorityQueue(exitsQueues[_token]);
+
+    // Iterate while the queue is not empty.
+    while (exitQueue.currentSize() > 0) {
+      (exitableAt, utxoPos) = getNextExit(_token);
+
+      // Check if this exit has finished its challenge period.
+      if (exitableAt > block.timestamp) {
+        return;
+      }
+
+      // get withdraw block
+      PlasmaExit memory currentExit = exits[utxoPos];
+
+      // process if NFT exists
+      // If an exit was successfully challenged, owner would be address(0).
+      address exitOwner = ExitNFT(exitNFTContract).ownerOf(utxoPos);
+      if (exitOwner != address(0)) {
+        if (_token == wethToken) {
+          // transfer ETH to token owner if `rootToken` is `wethToken`
+          WETH(wethToken).withdraw(currentExit.amount, exitOwner);
+        } else {
+          // transfer tokens to current contract
+          ERC20(_token).transfer(exitOwner, currentExit.amount);
+        }
+
+        // broadcast withdraw events
+        emit Withdraw(exitOwner, _token, currentExit.amount);
+
+        // Delete owner but keep amount to prevent another exit from the same UTXO.
+        // delete exits[utxoPos].owner;
+
+        // burn NFT
+        ExitNFT(exitNFTContract).burn(exitOwner, utxoPos);
+      }
+
+      // exit queue
+      exitQueue.delMin();
+    }
+  }
+
+  // withdraw tokens
+  function withdraw(
+    uint256 headerNumber,
+    bytes headerProof,
+
+    uint256 blockNumber,
+    uint256 blockTime,
+    bytes32 txRoot,
+    bytes32 receiptRoot,
+    bytes path,
+
+    bytes txBytes,
+    bytes txProof,
+
+    bytes receiptBytes,
+    bytes receiptProof
+  ) public {
+    address rootToken;
+    uint256 receiptAmount;
+
+    (rootToken, receiptAmount) = _processWithdrawReceipt(
+      receiptBytes,
+      path,
+      receiptProof,
+      receiptRoot,
+      msg.sender
+    );
+
+    // process withdraw tx
+    _processWithdrawTx(
+      txBytes,
+      path,
+      txProof,
+      txRoot,
+
+      rootToken,
+      receiptAmount,
+
+      msg.sender
+    );
+
+    // withdraw
+    _withdraw(
```

Comment:
1. This `_withdraw()` function is called by both burnProofWithdrawal-logic and exitWithdrawal-logic
2. This function appends withdrawal to priorityQueue finally
3. burnProofWithdrawal must call `WETH.withdrawal` or `ERC20.transfer` rather than `_withdraw`

Is this statement correct? @jdkanani

Reply:
No. `burnProofWithdrawal` will also call `_withdraw()` as operator can just create coins out of thin air, then burn it and create checkpoint without providing data - data unavailability issue.

So, if you burn then you can reuse your account. But if you don't burn and start exit without it, anyone can come and challenge if you reuse your account.

Reuse account == make transaction on Matic chain after exit

Reply:
1. 
> as operator can just create coins out of thin air, then burn it and create checkpoint without providing data - data unavailability issue.

make sense :)

2. 
> if you don't burn and start exit without it, anyone can come and challenge if you reuse your account.

Additionally anyone can cause state transition (e.g. balance addition) and the exit will be malicious.
So this exit function must be either "burnt exit" or "last resort"

Reply:
Yes!